### ISSUE TITLE
[bug] AptTool's use of apt-add-repository may update the apt index before the repository's key is added

### ISSUE THREAD

**Issue by hannahwhy** [conan-io/conan#8857]
### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 20.04
  * Compiler+version: gcc 9.3.0
  * Conan version: 1.35.1
  * Python version: Python 3.8.5

### Steps to reproduce (Include if Applicable)

In a `system_requirements` method, run

```python
if os_info.linux_distro == "ubuntu":
    apt = SystemPackageTool(tool=AptTool())
    apt_key = "https://apt.llvm.org/llvm-snapshot.gpg.key"

    if os_info.os_version == "20.04":
	apt.add_repository("'deb http://apt.llvm.org/focal/ llvm-toolchain-focal main'", apt_key)
``` 
On Ubuntu 20.04, `apt-add-repository` (which is invoked by `apt.add_repository`) defaults to running `apt update` after adding the repository line to apt's sources.  However, for a repository like `apt.llvm.org` whose signing key must usually be installed on the system, this will fail.

This is because `AptTool` installs the repository key _after_ running `apt-add-repository`.

### Logs (Executed commands with output) (Include/Attach if Applicable)

<details>
<summary>Log of a `conan install` of a Conanfile containing the above `system_requirements` code</summary>

```
recipes/toolchain-clang# conan create . toolchain-clang/11.0@
Exporting package recipe
toolchain-clang/11.0: The stored package has not changed
toolchain-clang/11.0: Exported revision: 29c135f3c45a6d6410c47bd6ecd52928
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=9
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

toolchain-clang/11.0: Forced build from source
Installing package: toolchain-clang/11.0
Requirements
    toolchain-clang/11.0 from local cache - Cache
Packages
    toolchain-clang/11.0:4db1be536558d833e52e862fd84d64d75c2b3656 - Build

Installing (downloading, building) binaries...
Running: sudo apt-add-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal main'
Hit:1 http://security.ubuntu.com/ubuntu focal-security InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease                            
Hit:4 http://archive.ubuntu.com/ubuntu focal-updates InRelease                    
Hit:5 http://archive.ubuntu.com/ubuntu focal-backports InRelease
Get:3 https://apt.llvm.org/focal llvm-toolchain-focal InRelease [5526 B]
Err:3 https://apt.llvm.org/focal llvm-toolchain-focal InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 15CF4D18AF4F7421
Reading package lists... Done
W: GPG error: https://apt.llvm.org/focal llvm-toolchain-focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 15CF4D18AF4F7421
E: The repository 'http://apt.llvm.org/focal llvm-toolchain-focal InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
toolchain-clang/11.0: ERROR: while executing system_requirements(): Command 'sudo apt-add-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal main'' failed
ERROR: Error in system requirements
```

</details>

**Comment by hannahwhy** [comment]
I have a proposed fix in https://github.com/yipdw/conan/commit/dcb45f6294f9f8baf07d4dc9fce44ee337fa554b, which involves moving the `apt-key add` invocation above the `apt-add-repository` invocation.  This ends up running `apt update` twice (because AptTool's `update()` is still run), but should be compatible with all released versions of `apt-add-repository`.

It is also possible to run `apt-add-repository --no-update`, but I'm not sure when that option was added.

**Comment by memsharded** [comment]
> I have a proposed fix in yipdw@dcb45f6, which involves moving the apt-key add invocation above the apt-add-repository invocation. This ends up running apt update twice (because AptTool's update() is still run), but should be compatible with all released versions of apt-add-repository.

If I understand correctly, this would be 0 risk, but introduce some extra slowness running an extra update?. Maybe you want to do a PR with that change, and we could fire that conversation. Some people might not like that extra time, but we can try.

> It is also possible to run apt-add-repository --no-update, but I'm not sure when that option was added.

Yes, this could be risky, and breaking to users with older distros, if --no-update was introduced recently. We would need extra validation for this to be considered.

**Comment by hannahwhy** [comment]
Sure thing, I'll open a PR with https://github.com/yipdw/conan/commit/dcb45f6 and we can try it out.

**Comment by hannahwhy** [comment]
PR #8861 opened.

**Comment by memsharded** [comment]
Will be released in 1.37, many thanks!
